packages/plugin-data/src/commands/force/data/soql/query.ts

```diff
@@ -0,0 +1,140 @@
+
+export class DataSoqlQueryCommand extends SfdxCommand {
+  public static readonly description = messages.getMessage('description');
+  public static readonly longDescription = messages.getMessage('longDescription');
+  public static readonly requiresProject = false;
+  public static readonly requiresUsername = true;
+  public static readonly examples = messages.getMessage('examples').split(os.EOL);
+
+  public static readonly flagsConfig: FlagsConfig = {
+    query: flags.string({
+      char: 'q',
+      required: true,
+      hidden: false,
+      description: messages.getMessage('queryToExecute'),
+      longDescription: messages.getMessage('queryLongDescription'),
+    }),
+    usetoolingapi: flags.boolean({
+      char: 't',
+      required: false,
+      hidden: false,
+      default: false,
+      description: messages.getMessage('queryToolingDescription'),
+      longDescription: messages.getMessage('queryToolingLongDescription'),
+    }),
+    resultformat: flags.enum({
+      char: 'r',
+      description: messages.getMessage('resultFormatDescription'),
+      longDescription: messages.getMessage('resultFormatLongDescription'),
+      options: ['human', 'csv', 'json'],
+      default: 'human',
+    }),
+  };
+
+  /**
+   * Define display function that will produce the desired output based on flag resultformat selection
+   *
+   * @protected
+   */
+  protected static readonly result: SfdxResult = {
+    display(): void {
+      const results = asPlainObject(this.data) as DataSoqlQueryResult;
+      let reporter;
+      switch (results.resultFormat as keyof typeof FormatTypes) {
+        case 'human':
+          reporter = new HumanReporter(results, results.columns, this.ux, results.logger);
+          break;
+        case 'json':
+          reporter = new JsonReporter(results, results.columns, this.ux, results.logger);
+          break;
+        case 'csv':
+          reporter = new CsvReporter(results, results.columns, this.ux, results.logger);
+          break;
+        default:
+          throw new Error(`result format is invalid: ${results.resultFormat}`);
+      }
+      // delegate to selected reporter
+      reporter.display();
+    },
+  };
+
+  // Overrides SfdxCommand.  This is ensured since requiresUsername == true
+  protected org!: Org;
+
+  /**
+   * Command run implementation
+   *
+   * Returns either a DataSoqlQueryResult or a SfdxResult.
+   * When the user is using global '--json' flag an instance of SfdxResult si returned.
+   * This is necessary since '--json' flag reports results in the form of SfdxResult
+   * and bypasses the definition of start result. The goal is to have the output
+   * from '--json' and '--resulformat json' be the same.
+   *
+   * The DataSoqlQueryResult is necessary to communicate user selections to the reporters.
+   * The 'this' object available during display() function does not include user input to
+   * the command, which are necessary for reporter selection.
+   *
+   */
```

Comment:
Maybe we should modify SfdxResult to support the usecase so we don't have to modify the return of `run`?  I think it was just an oversight on my part when I wrote that long ago.

Reply:
@shetzel, I agree, but should that be handled in a different user story?

Reply:
Yeah, that's a good idea.